src/qiskit_qasm3_import/expression.py

```diff
@@ -55,32 +69,47 @@ class ValueResolver(QASMVisitor):
     These resolvers are intended to be short-lived.  Their instance state is the symbols that are
     available to them.  There is no per-run state tracked."""
 
-    __slots__ = ("symbols",)
+    __slots__ = ("context",)
 
     # pylint: disable=no-self-use
 
-    def __init__(self, symbols: Mapping[str, Symbol]):
-        self.symbols = symbols
+    def __init__(self, context: "State"):
+        self.context = context
 
     def resolve(self, node: ast.Expression) -> Tuple[Any, types.Type]:
         """The entry point to the resolver, resolving the AST node into a 2-tuple of a relevant
         Qiskit type, and the :class:`.Type` that it is an instance of."""
+
         return self.visit(node)
 
-    def visit(self, node: ast.QASMNode, context: None = None) -> Tuple[Any, types.Type]:
-        value, type = super().visit(node, context)
+    # pylint: disable=arguments-differ
+    def visit(self, node: ast.QASMNode) -> Tuple[Any, types.Type]:
+        value, type = super().visit(node)
         if isinstance(type, types.Error):
             raise_from_node(node, "type error")
         return value, type
 
-    def generic_visit(self, node: ast.QASMNode, context: None = None):
+    # pylint: disable=arguments-differ
+    def generic_visit(self, node: ast.QASMNode):
         raise_from_node(node, f"'{node.__class__.__name__}' cannot be resolved into a Qiskit value")
 
     def visit_Identifier(self, node: ast.Identifier):
         name = node.name
-        if name not in self.symbols:
-            raise_from_node(node, f"name '{name}' is not defined in this scope")
-        symbol = self.symbols[name]
+        if name not in self.context.symbol_table:
+            if is_physical(name):  # Physical qubits are not declared.
+                if self.context.scope == Scope.GATE:
+                    raise_from_node(
+                        node,
+                        f"Illegal qubit reference '{name}'. References to hardware qubits not allowed in gate definitions.",
+                    )
+                self.context.addressing_mode.set_physical_mode(node)
+                self.context.circuit.add_bits([bit := Qubit()])
+                symbol = Symbol(name, bit, types.HardwareQubit(), Scope.GLOBAL, node)
+                self.context.symbol_table[name] = symbol
```

Comment:
This needs to be added to the global scope.  We probably want to more properly encapsulate the symbol table into a class that can hold its own stack at this point, and have `SymbolTable.get` and `SymbolTable.set_local` and `SymbolTable.set_global` methods on it.

Reply:
I favor encapsulating the symbol table.

All of the state is in `State`. A new `State` is created for locate and gate scopes  (`FUNCTION` and `CALIBRATION` are currently unused). These structures are temporary and no information is propagated explicitly. Rather it is controlled by either copying or referring to the data in the parent `State`. A global symbol table in `State` might work with this scheme. I didn't think through the details. But encapsulated access to symbol tables might take care of the complexity of handling two tables. For example you can query a symbol name, or restrict to only local or global symbols.

Reply:
I am resolving this conversation. The design of `SymbolTable` is discussed in another conversation in this review.